Add basic reconciliation metrics to controllers

pmorie opened the pull request:
#### What type of PR is this?

This supercedes https://github.com/kubernetes/kubernetes/pull/100709

/kind feature

#### What this PR does / why we need it:

This adds a histogram metric to various controllers, `controller_reconcile_time_seconds`.

Example output:
```
# HELP controller_reconcile_time_seconds [ALPHA] Duration of a sync of a controller's resource
# TYPE controller_reconcile_time_seconds histogram
controller_reconcile_time_seconds_bucket{controller="statefulset",status="error",le="0.001"} 0
controller_reconcile_time_seconds_bucket{controller="statefulset",status="error",le="0.002"} 0
controller_reconcile_time_seconds_bucket{controller="statefulset",status="error",le="0.004"} 0
controller_reconcile_time_seconds_bucket{controller="statefulset",status="error",le="0.008"} 0
controller_reconcile_time_seconds_bucket{controller="statefulset",status="error",le="0.016"} 0
controller_reconcile_time_seconds_bucket{controller="statefulset",status="error",le="0.032"} 6
controller_reconcile_time_seconds_bucket{controller="statefulset",status="error",le="0.064"} 8
controller_reconcile_time_seconds_bucket{controller="statefulset",status="error",le="0.128"} 16
controller_reconcile_time_seconds_bucket{controller="statefulset",status="error",le="0.256"} 37
controller_reconcile_time_seconds_bucket{controller="statefulset",status="error",le="0.512"} 47
controller_reconcile_time_seconds_bucket{controller="statefulset",status="error",le="1.024"} 49
controller_reconcile_time_seconds_bucket{controller="statefulset",status="error",le="2.048"} 49
controller_reconcile_time_seconds_bucket{controller="statefulset",status="error",le="4.096"} 49
controller_reconcile_time_seconds_bucket{controller="statefulset",status="error",le="8.192"} 49
controller_reconcile_time_seconds_bucket{controller="statefulset",status="error",le="16.384"} 49
controller_reconcile_time_seconds_bucket{controller="statefulset",status="error",le="+Inf"} 49
controller_reconcile_time_seconds_sum{controller="statefulset",status="error"} 9.560064066000002
controller_reconcile_time_seconds_count{controller="statefulset",status="error"} 49
controller_reconcile_time_seconds_bucket{controller="statefulset",status="success",le="0.001"} 9
controller_reconcile_time_seconds_bucket{controller="statefulset",status="success",le="0.002"} 62
controller_reconcile_time_seconds_bucket{controller="statefulset",status="success",le="0.004"} 222
controller_reconcile_time_seconds_bucket{controller="statefulset",status="success",le="0.008"} 266
controller_reconcile_time_seconds_bucket{controller="statefulset",status="success",le="0.016"} 280
controller_reconcile_time_seconds_bucket{controller="statefulset",status="success",le="0.032"} 307
controller_reconcile_time_seconds_bucket{controller="statefulset",status="success",le="0.064"} 328
controller_reconcile_time_seconds_bucket{controller="statefulset",status="success",le="0.128"} 341
controller_reconcile_time_seconds_bucket{controller="statefulset",status="success",le="0.256"} 347
controller_reconcile_time_seconds_bucket{controller="statefulset",status="success",le="0.512"} 347
controller_reconcile_time_seconds_bucket{controller="statefulset",status="success",le="1.024"} 352
controller_reconcile_time_seconds_bucket{controller="statefulset",status="success",le="2.048"} 352
controller_reconcile_time_seconds_bucket{controller="statefulset",status="success",le="4.096"} 352
controller_reconcile_time_seconds_bucket{controller="statefulset",status="success",le="8.192"} 352
controller_reconcile_time_seconds_bucket{controller="statefulset",status="success",le="16.384"} 352
controller_reconcile_time_seconds_bucket{controller="statefulset",status="success",le="+Inf"} 352
controller_reconcile_time_seconds_sum{controller="statefulset",status="success"} 8.841072798
controller_reconcile_time_seconds_count{controller="statefulset",status="success"} 352
```

#### Does this PR introduce a user-facing change?
```release-note
adds reconciliation metrics for kube-controller-manager.
```

pmorie commented:
@cheftako I'll address these in the next couple days:

- https://github.com/kubernetes/kubernetes/pull/100709#discussion_r690077830
- https://github.com/kubernetes/kubernetes/pull/100709#discussion_r690081589

roycaihw commented:
/remove-sig api-machinery

alculquicondor commented on pkg/controller/job/job_controller.go:
```diff
@@ -170,7 +170,7 @@ func NewController(podInformer coreinformers.PodInformer, jobInformer batchinfor
 	jm.updateStatusHandler = jm.updateJobStatus
 	jm.patchJobHandler = jm.patchJob
-	jm.syncHandler = jm.syncJob
+	jm.syncHandler = controller.SyncAndRecordWithBool("job", jm.syncJob)
```
We already have a metric specific for the job controller https://github.com/kubernetes/kubernetes/blob/master/pkg/controller/job/metrics/metrics.go#L43

Is it worth having 2 metrics? I think we should standardize the name, but leave it up to the controller to add the labels they might need.

pmorie replied:
I am not sure what you are suggesting, would you elaborate?

alculquicondor replied:
I am suggesting that this change is reverted for the job controller, as it already defines a metric with the same intent.

But we can standardize the name so that it's a consistent experience for people that need these metrics.

pmorie replied:
got it, thanks!

alculquicondor replied:
cc @soltysh for thoughts.

alculquicondor replied:
@logicalhan wdyt?

I don't think it's good if we have 2 metrics that provide very similar output

logicalhan replied:
I like having the generic controller metrics, I think it is generally a step in the right direction. Are you saying there are already similar metrics for the scheduler?

alculquicondor replied:
no, for the job controller. There is already a metric that has the same intent, but has domain-specific labels
https://github.com/kubernetes/kubernetes/blob/6f896dec4f45ffb82491bc9ce2393e7452886562/pkg/controller/job/metrics/metrics.go#L30-L44

alculquicondor replied:
how much structure can we put in "additional_data"?

logicalhan replied:
I was thinking it could just be a catch-all additional label, so for jobs it would take the place of `completion_mode`.

logicalhan replied:
I mean what's the point of a generic controller metric if you are going to special case away from it every time...

pmorie replied:
i'm +1 on adding `"action"` and `"additional_data"`. I would see usage of `"additional_data"` in the job controller having values like `"completion_mode=NonIndexed"`. That said, I am not deeply nuanced in my understanding of what values metric labels are expected to have; is what I described valid?

pmorie replied:
i guess an alternative to `additional_data` might be to have specific metrics for additional information specific to individual controllers. Again, I'm not deeply nuanced on how things like single table inheritance apply to metrics, just 'thinking aloud'.

alculquicondor replied:
What is the final decision here? Are we proceeding with duplicate metrics?

alculquicondor commented on pkg/controller/metrics.go:
```diff
@@ -0,0 +1,145 @@
+	syncDuration := time.Since(startTime).Seconds()
+
+	ReconciliationDurations.WithLabelValues(controllerName, status).Observe(syncDuration)
+	klog.V(4).Infof("Controller %s finished syncing in %.3f seconds with status %s",
```
Use structured logging.

pmorie replied:
Ack, i will make that change

alculquicondor commented on pkg/controller/metrics.go:
```diff
@@ -0,0 +1,145 @@
+limitations under the License.
+*/
+
+package controller
```
probably worth being in its own package with an OWNERS file owned by sig-instrumentation.

pmorie replied:
Ack

logicalhan replied:
any update on this?

pmorie replied:
@logicalhan I've just pushed some updates, apologies for the delay.

logicalhan replied:
> @logicalhan I've just pushed some updates, apologies for the delay.

We actually put these sorts of metrics into component-base, like so: https://github.com/kubernetes/kubernetes/tree/master/staging/src/k8s.io/component-base/metrics/prometheus

pmorie replied:
@logicalhan: should I do component-base/metrics/controller?

logicalhan replied:
no it should be `/component-base/metrics/prometheus/controller`

pmorie replied:
@logicalhan got it, i'll do that in my next work session. any feedback on the the metrics_test.go file in this PR?

dims commented:
/assign @alculquicondor

dashpole commented:
/assign @logicalhan 
to review for sig-instrumentation

dashpole commented:
/triage accepted

pmorie commented:
new changes coming for this one shortly!

pmorie commented:
I've extracted the metrics helpers into their own package and added tests.

Still need to:

- determine how to resolve job controller or standardize metric name
- ensure all controllers within KCM are covered

logicalhan commented on staging/src/k8s.io/component-base/metrics/prometheus/controller/metrics.go:
```diff
@@ -0,0 +1,151 @@
+}
+
+// SyncAndRecordWithBool wraps syncFn with a function, recording its result and duration as a metric.
+func SyncAndRecordWithBool(
```
nit: prefer `HasSyncedAndRecorded`

logicalhan replied:
the `WithBool` reads poorly...

pmorie replied:
thanks for the review, i'll incorporate these in my next working session on this tomorrow.

pmorie replied:
Let's make sure we're on same page before I do the renames:

- `SyncAndRecordWithBool` -> `HasSyncedAndRecordedWithCtx`
- `SyncAndRecordWithBoolNoCtx` -> `HasSyncedAndRecorded`
- `RunSyncAndRecordWithBool` -> `RunHasSyncedAndRecordedWithCtx`
- `RunSyncAndRecordWithBoolNoCtx` -> `RunHasSyncedAndRecorded`
- `SyncAndRecord` -> `SyncAndRecordWithCtx`
- `SyncAndRecordNoCtx` -> `SyncAndRecord`
- `RunSyncAndRecord` -> `RunSyncAndRecordWithCtx`
- `RunSyncAndRecordNoCtx` -> `RunSyncAndRecord`

LMK

pmorie replied:
@logicalhan you good with these renames ^ ?

logicalhan commented on staging/src/k8s.io/component-base/metrics/prometheus/controller/metrics.go:
```diff
@@ -0,0 +1,151 @@
+}
+
+// SyncAndRecordWithBoolNoCtx wraps syncFn with a function, recording its result and duration as a metric.
+func SyncAndRecordWithBoolNoCtx(
```
I would prefer `HasSyncedAndRecorded` and `HasSyncedAndRecordedWithCtx` rather than a `NoCtx` variant.

logicalhan commented:
(Also you are failing dep checks and verify tests)

pmorie commented:
I had some unexpected work demands today; will finish method renames tomorrow. Lint and verify jobs should be fixed now.

pmorie commented:
oops, the build failures are expected. will fix in the am

logicalhan commented on staging/src/k8s.io/component-base/metrics/prometheus/controller/metrics.go:
```diff
@@ -0,0 +1,151 @@
+}
+
+// RunSyncAndRecord runs syncFn with the given context and key, recording the reconcile time and results
+func RunSyncAndRecord(
```
this one should be WithCtx

logicalhan commented on staging/src/k8s.io/component-base/metrics/prometheus/controller/metrics.go:
```diff
@@ -0,0 +1,151 @@
+}
+
+// RunSyncAndRecordNoCtx runs syncFn with key, recording the reconcile time and results
+func RunSyncAndRecordNoCtx(
```
and the NoCtx here can be omitted

pmorie replied:
Yep, ack. I think we're on the same page now. Thanks!

alculquicondor commented on staging/src/k8s.io/component-base/metrics/prometheus/controller/metrics.go:
```diff
@@ -0,0 +1,151 @@
+			StabilityLevel: metrics.ALPHA,
+			Buckets:        metrics.ExponentialBuckets(0.001, 2, 15),
+		},
+		[]string{"controller", "status"},
```
Sorry if this has been discussed before, but would it make sense to have an instance of the metric for each subsystem, instead of having the controller as a label?

pmorie commented:
/retest

pmorie commented:
/test pull-kubernetes-e2e-gce-ubuntu-containerd

mrbobbytables commented:
/easycla

pmorie commented:
/easycla

dims commented:
Please see https://github.com/kubernetes/community/blob/master/CLA.md#the-contributor-license-agreement on how to sign the CLA.

If you have already signed, please rebase commits in this PR to latest master and reopen the PR. This will trigger the newer EasyCLA bot to run and clear the "cncf-cla: no" label. We cannot merge any PRs if the CLA is not signed.

Thanks!

pmorie commented:
I see now that apparently my current work email, which is associated with the commit in this PR, wasn't verified, I wonder if that could be it.

/easycla

pmorie commented:
I see now what I did, should be fixed up shortly

pmorie committed:
Add basic reconciliation metrics to controllers

pmorie commented:
/test pull-kubernetes-node-e2e-containerd

pmorie commented:
/test pull-kubernetes-integration

pmorie commented:
/test pull-kubernetes-integration

thehackercat commented:
any progress?

ritazh commented:
/remove-sig auth
Feel free to add us back if needed.
